### ISSUE TITLE
Minify and concatenate dashboard and client scripts

### ISSUE THREAD

**Issue by gregpabian** [benderjs/benderjs#186]
Some scripts loaded for the dashboard and the client page are not minified nor concatenated. We can:

1. Do this on-the-fly when the server starts, or
2. Introduce a builder to do this once before every release and use the minified/concatenated version.

**Comment by Reinmar** [comment]
+1 for 1st. 2nd solution is more painful to live with.

**Comment by fredck** [comment]
I think "1" is the only option, after all plugins stuff should be also merged in, shouldn't it?

**Comment by gregpabian** [comment]
Merging plugin resources might be a bit tricky - some of the plugins may load their resources conditionally, e.g. benderjs-sinon loads an additional script for IE only if some specific conditions are met. I'm not saying it's impossible, we just have to work out a smart solution for this, instead of merging all the stuff that sits in plugin's `files` array.

**Comment by fredck** [comment]
I think we miss API in Bender for this... to programatically set JavaScript and CSS to be loaded by Bender. This would be used by Bender's core as well as plugins.

**Comment by Reinmar** [comment]
I don't think that Sinon, Jasmin, etc should be minified. It will make debugging harder. Don't forget that tests are used for development so libraries that are part of the development process should not be minified. Bender's dashboard is a different case, because you don't need to see its code usually.

**Comment by fredck** [comment]
> I don't think that Sinon, Jasmin, etc should be minified.

I think this should be the default behavior instead... it is useless to have the framework files debuggable by default. Lack of trust on them is not a good excuse for this. Only tests and application files should be delivered untouched.

For debugging purposes, I would enable a Bender configuration that would inject the scripts in separate files without minification.

**Comment by gregpabian** [comment]
I'd go with a configuration option to disable the minification too.

**Comment by Reinmar** [comment]
> Only tests and application files should be delivered untouched.

I strongly disagree. It often happens that external libs' calls are part of a stack trace. We are talking here about YUI Test, Sinon, CKEditor testing tools (?!!!), etc. By minifying all them we will make debugging harder for no good reason. I will need to open Bender's config, change the option, save it, restart Bender, debug, then revert everything. Absolute productivity killer. And at the same time... what is the added value of minifying them? Minor performance improvement.

Really, I could only accept minifying files that I don't care while developing. If you decide to implement it as an option I will have it always disabled.

**Comment by fredck** [comment]
> It often happens that external libs' calls are part of a stack trace.

And what's the debugging problem with this? You would just pass through minified code, knowing that you can ignore that because it is clear that is not the code that interests you on debugging. This would be even clear in the stack trace. Or you just want to have nice function names there? 

Maybe I'm missing something, but I still don't see the benefit of not minifying them.

> what is the added value of minifying them? Minor performance improvement.

This may be special impact on the test run, because each test page loaded right now loads lots of resources which could be merged.

Additionally, earlier or later we gonna have people running remote tests and these kinds of improvements are welcome.

> If you decide to implement it as an option I will have it always disabled.

That's a good thing as well... one can disable it if finds it necessary.

**Comment by Reinmar** [comment]
> This may be special impact on the test run, because each test page loaded right now loads lots of resources which could be merged.
> 
> Additionally, earlier or later we gonna have people running remote tests and these kinds of improvements are welcome.

When you run all the tests at once you have all the files cached. So most of the advantage of concatenation is lost. Concatenation and minification matter most when having clear or disabled cache.

> And what's the debugging problem with this? You would just pass through minified code, knowing that you can ignore that because it is clear that is not the code that interests you on debugging. This would be even clear in the stack trace. Or you just want to have nice function names there?
> 
> Maybe I'm missing something, but I still don't see the benefit of not minifying them.

I'm not 100% sure now that CKEditor testing tools would be minified too. But assuming that they would, I could not clearly debug what's happening in there and we're talking about plenty of code. I couldn't also easily place there debugger statements or console logs or simply follow what's happening inside and often these functions make lots of calls to CKEditor APIs so they are part of the problem themselves. It's a smaller problem with YUI Test or Sinon of course, but occasionally you may want to see inside them too. And most importantly - you want to be able to associate specific calls in a stack trace with them.

So again - little added value due to caching when running full tests suite and possible problems when minified too much. That's why I wouldn't go crazy with minification.

PS. No one mentioned source maps... and actually they would help but not for all problems. Still we're debugging on old browsers and we sometimes modify testing tools on the fly.

**Comment by fredck** [comment]
The requirement you described looks like a very uncommon task. Not the thing that we would do every day.

I would proceed with this and make it configurable, simply. We can call theories about performance and usability here but until we'll not see it for real it would be hard for us to surely state anything. Making it configurable would solve the individual requirements and we can at any moment change the defaults if the decision we take will reveal to be strongly wrong. It is a welcome feature anyway.

**Comment by jodator** [comment]
@Reinmar Does sourceMap would be a good option?

**Comment by Reinmar** [comment]
> @Reinmar Does sourceMap would be a good option?

As I wrote - they'd solve part of the problems. Mostly following stack traces in modern browsers. But there you have old IEs and sometimes you want to log some intermediate values in your testing tools.

> The requirement you described looks like a very uncommon task. Not the thing that we would do every day.

I wouldn't mention this problem if I haven't encounter it often enough to make a difference. It's enough that it happens once a week to piss a developer off.

> I would proceed with this and make it configurable, simply. We can call theories about performance and usability here but until we'll not see it for real it would be hard for us to surely state anything. Making it configurable would solve the individual requirements and we can at any moment change the defaults if the decision we take will reveal to be strongly wrong. It is a welcome feature anyway.

But I don't say that adding minification is not worth the effort. I'm saying that we should do it in a reasonable way. If I'll need to disable it will be a wasted time (for me at least).

**Comment by Reinmar** [comment]
If you don't see my point, then we could at least find a middle ground which for me could be to have a way to exclude CKEditor testing tools from minification. Libraries like YUI Test or Sinon are less important.

BTW. There's one more argument - CKEditor plugin for Bender is part of ckeditor-dev so when you switch branches it should be refreshed (in ckeditor-dev it changes pretty often). Actually - this is the most critical issue.

**Comment by fredck** [comment]
Maybe the API could provide a way for the plugin to say: "hey, include this, but don't merge and minify it".

**Comment by jodator** [comment]
> As I wrote - they'd solve part of the problems. Mostly following stack traces in modern browsers. But there you have old IEs and sometimes you want to log some intermediate values in your testing tools.

@Reinmar sorry I've just scanned briefly through you replies and forgot about old IE's :D 

> Maybe the API could provide a way for the plugin to say: "hey, include this, but don't merge and minify it".

Maybe one config option? Enable/Disable whole (simpler API)
